### PR TITLE
TDL-17698: Dict based to class based refactoring

### PR DESCRIPTION
**Pull request by harshpatel4crest**
# Description of change
TDL-17698: Dict based to class based refactoring
- Updated code to class-based.
- Added sync logic from sync.py to `GoogleSheets` class.
- No difference is found in the catalog and state file.
- Integration tests passing locally.

# Manual QA steps
 - 
 
# Risks
 - 
 
# Rollback steps
 - revert this branch

### CODE REVIEW

File: tap_google_sheets/streams.py [hunk 1]
```diff
@@ -1,74 +1,506 @@
...
+    stream = catalog.get_stream(stream_name)
+    schema = stream.schema.to_dict()
+    try:
+        singer.write_schema(stream_name, schema, stream.key_properties)
+        LOGGER.info('Writing schema for: {}'.format(stream_name))
+    except OSError as err:
+        LOGGER.info('OS Error writing schema for: {}'.format(stream_name))
+        raise err
+
+def write_record(stream_name, record, time_extracted, version=None):
```
Comment (dbshah1212) [inline tap_google_sheets/streams.py:43]: Add function comment, why do we need to check the version?

File: tap_google_sheets/streams.py [hunk 2]
```diff
@@ -1,74 +1,506 @@
...
+        else:
+            singer.messages.write_record(
+                stream_name=stream_name,
+                record=record,
+                time_extracted=time_extracted)
+    except OSError as err:
+        LOGGER.info('OS Error writing record for: {}'.format(stream_name))
+        raise err
+
+def get_bookmark(state, stream, default):
```
Comment (dbshah1212) [inline tap_google_sheets/streams.py:61]: Add function comment to all functions.

**Review by dbshah1212 — CHANGES_REQUESTED** [review]
Provide function comments

File: tap_google_sheets/discover.py [hunk 3]
```diff
@@ -1,26 +1,40 @@
...
+
+            key_properties = None
+
+            # get primary key for the stream
+            for mdt in mdata:
+                table_key_properties = mdt.get('metadata', {}).get('table-key-properties')
+                if table_key_properties:
+                    key_properties = table_key_properties
+
+            if not STREAMS.get(stream_name):
```
[the remark this answers is not in the document]
Comment (harshpatel4crest) [reply tap_google_sheets/discover.py:29]: Added comment.

File: tap_google_sheets/streams.py (hunk 1 above)
Comment (harshpatel4crest) [reply tap_google_sheets/streams.py:46]: Added comment.

File: tap_google_sheets/streams.py (hunk 2 above)
Comment (harshpatel4crest) [reply tap_google_sheets/streams.py:69]: Added comments.

File: tap_google_sheets/discover.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (dbshah1212) [reply tap_google_sheets/discover.py:29]: Why do we need to do this? is there a possibility to get a stream which is not available in STREAMS?

File: tap_google_sheets/streams.py [hunk 4]
```diff
@@ -1,74 +1,519 @@
...
+        raise err
+
+def write_record(stream_name, record, time_extracted, version=None):
+    """
+    Write records for the stream with extracted time
+    """
+    try:
+        # use "write_message" if version is found as
+        # "write_record" params does not contain "version"
+        if version:
```
Comment (dbshah1212) [inline tap_google_sheets/streams.py:53]: What is the difference between 2 calls? Will both return the same data?

File: tap_google_sheets/discover.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (harshpatel4crest) [reply tap_google_sheets/discover.py:29]: Yes, this is Tap is a mixture of dynamic and static streams. The sheets that we create in the spreadsheet, will be discovered dynamically.

File: tap_google_sheets/streams.py [hunk 5]
```diff
@@ -1,74 +1,519 @@
...
+            schema=schema,
+            key_properties=self.key_properties,
+            valid_replication_keys=self.replication_keys,
+            replication_method=self.replication_method
+        )
+        field_metadata[self.stream_name] = mdata
+
+        return schemas, field_metadata
+
+    # Transform/validate batch of records w/ schema and sent to target
```
Comment (dbshah1212) [inline tap_google_sheets/streams.py:153]: What w/ means?

Comment (dbshah1212) [reply tap_google_sheets/streams.py:153]: Please keep consistency in function comments, add below the def fun_name in all functions.

File: tap_google_sheets/streams.py (hunk 4 above)
Comment (harshpatel4crest) [reply tap_google_sheets/streams.py:53]: Yes, the `write_record` internally calls the `write_message` function. [Reference](https://github.com/singer-io/singer-python/blob/0c066de21111d8572425083b4a8792d193c80af1/singer/messages.py#L230)

File: tap_google_sheets/streams.py (hunk 5 above)
Comment (harshpatel4crest) [reply tap_google_sheets/streams.py]: Added comments below the function declaration.

**Review by dbshah1212 — APPROVED** [review]

**Review by kspeer825 — APPROVED** [review]

File: tap_google_sheets/schema.py [hunk 6]
```diff
@@ -240,16 +237,16 @@ def get_sheet_metadata(sheet, spreadsheet_id, client):
...
+    api = stream_obj.api
+    params = stream_obj.params
     sheet_title_encoded = urllib.parse.quote_plus(sheet_title)
     sheet_title_escaped = re.escape(sheet_title)
     # create querystring for preparing the request
     querystring = '&'.join(['%s=%s' % (key, value) for (key, value) in \
         params.items()]).replace('{sheet_title}', sheet_title_encoded)
     # create path for preparing the request
-    path = '{}?{}'.format(stream_metadata.get('path').replace('{spreadsheet_id}', \
+    path = '{}?{}'.format(stream_obj.path.replace('{spreadsheet_id}', \
```
[the remark this answers is not in the document]
Comment (harshpatel4crest) [reply tap_google_sheets/schema.py]: Created a function to return the path with query string.

File: tap_google_sheets/discover.py [hunk 7]
```diff
@@ -1,26 +1,43 @@
...
+            schema = Schema.from_dict(schema_dict)
+            mdata = field_metadata[stream_name]
+
+            key_properties = None
+
+            # get primary key for the stream
+            for mdt in mdata:
+                table_key_properties = mdt.get('metadata', {}).get('table-key-properties')
+                if table_key_properties:
+                    key_properties = table_key_properties
```
[the remark this answers is not in the document]
Comment (harshpatel4crest) [reply tap_google_sheets/discover.py]: Added this code change in the **if** condition block.

File: tap_google_sheets/transform.py [hunk 8]
```diff
@@ -0,0 +1,231 @@
...
+        return int(value)
+    elif isinstance(value, float):
+        return transform_sheet_decimal_data(value, sheet_title, col_name, col_letter, row_num, col_type)
+    else:
+        LOGGER.info('WARNING: POSSIBLE DATA TYPE ERROR: SHEET: {}, COL: {}, CELL: {}{}, TYPE: {}'.format(
+                sheet_title, col_name, col_letter, row_num, col_type))
+        return str(value)
+
+# return transformed column the values based on the datatype
+def get_column_value(value, sheet_title, col_name, col_letter, row_num, col_type, row):
```
Comment (KrisPersonal) [inline tap_google_sheets/transform.py:157]: I don't see row_num argument not being used. If not used why do we need the argument being passed

File: tap_google_sheets/transform.py [hunk 9]
```diff
@@ -0,0 +1,231 @@
...
+        except ValueError:
+            col_val = str(value)
+            LOGGER.info('WARNING: POSSIBLE DATA TYPE ERROR; SHEET: {}, COL: {}, CELL: {}{}, TYPE: {}'.format(
+                sheet_title, col_name, col_letter, row_num, col_type))
+        return col_val
+    else:
+        return str(value)
+
+# transform boolean values in the sheet
+def transform_sheet_boolean_data(value, sheet_title, col_name, col_letter, row_num, col_type, row):
```
Comment (KrisPersonal) [inline tap_google_sheets/transform.py:99]: Where is row_num being used in this function?

File: tap_google_sheets/transform.py [hunk 10]
```diff
@@ -0,0 +1,231 @@
...
+
+                    # get column value based on the type of the value
+                    col_val = get_column_value(value, sheet_title, col_name, col_letter, row_num, col_type, row)
+
+                    sheet_data_row_tf[col_name] = col_val
+                col_num = col_num + 1
+            # APPEND non-empty row
+            sheet_data_tf.append(sheet_data_row_tf)
+        row_num = row_num + 1
+    return sheet_data_tf, row_num
```
Comment (KrisPersonal) [inline tap_google_sheets/transform.py:231]: Did you guys pylint on this transform function? Some of the issues needs to be fixed for readability

**Review by KrisPersonal — COMMENTED** [review]
Please run pylint on schema.py and fix unused variable issue

**Review by KrisPersonal — COMMENTED** [review]
Run pylint on all .py files and fix obvious issues like variable not being used. Let us clean-up the code more

File: tap_google_sheets/transform.py (hunk 8 above)
Comment (harshpatel4crest) [reply tap_google_sheets/transform.py:157]: The **row_num** is being used in this block of code.

File: tap_google_sheets/transform.py (hunk 9 above)
Comment (harshpatel4crest) [reply tap_google_sheets/transform.py]: Removed row_num argument from the function.

**Comment by harshpatel4crest** [comment]
@KrisPersonal Pylint is [already running](https://github.com/singer-io/tap-google-sheets/blob/master/.circleci/config.yml#L21) on this Tap as `pylint tap_google_sheets -d C,R,W`.

File: tap_google_sheets/transform.py (hunk 10 above)
Comment (harshpatel4crest) [reply tap_google_sheets/transform.py:231]: Pylint is running on **tap_google_sheets** folder

**Review by KrisPersonal — APPROVED** [review]

**Pull request merged**